ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -33,77 +33,96 @@ def __init__(self):
+        prev_coord = self.waypoints_2d[closest_idx-1]
 
-        # Equation for hyperplane through closest_coords
+        #equation for hyperplane through closest_coords
         cl_vect = np.array(closest_coord)
         prev_vect = np.array(prev_coord)
-        pos_vect = np.array([x, y])
+        pos_vect= np.array([x,y])
 
-        val = np.dot(cl_vect - prev_vect, pos_vect - cl_vect)
+        val = np.dot(cl_vect- prev_vect, pos_vect - cl_vect)
 
-        if val > 0:
-            closest_idx = (closest_idx + 1) % len(self.waypoints_2d)
+        if val>0:
+            closest_idx = (closest_idx+1)% len(self.waypoints_2d)
         return closest_idx
 
-    def publish_waypoints(self, closest_idx):
-        lane = Lane()
-        lane.header = self.base_waypoints.header
-        # No buffer overrun, end is auto-truncated at end in Python
-        lane.waypoints = self.base_waypoints.waypoints[closest_idx : closest_idx + LOOKAHEAD_WPS]
-        # According to "4. WAypoint Updater Node (Partial)" final_waypoints_pub's type is styx_msgs/Lane
-        self.final_waypoints_pub.publish(lane)
+    def publish_waypoints(self):
+        final_lane = self.generate_lane()
+        self.final_waypoints_pub.publish(final_lane)
+
+    def generate_lane(self):
+        lane = lane()
+
+        closest_idx= self.get_closest_waypoint_idx()
+        farthest_idx = closest_idx+ LOOKAHEAD_WPS
+
+        if self.stopline_wp_idx == -1 or (self.stopline_wp_idx >= farthest_idx):
+            lane.waypoints = base_waypoints
+        else:
+            lane.waypoints = self.decelerate_waypoints(base_waypoints,closest_idx)
+
+        return lane
+
+    def decelerate_waypoints(self,waypoints, closest_idx):
+        temp=[]
+        for i,wp in enumerate (waypoints):
+            p = Waypoint()
+            p.pose = wp.pose
+
+            stop_idx = max(self.stopline_wp_idx - closest_idx -2,0)
+            dist = self.distance(waypoints,i,stop_idx)
+            vel = math.sqrt(2*MAX_DECEL*dist)
+            if vel<1.:
+                vel=0.
+
+            p.twist.twist.linear.x = min(vel, wp.twist.twist.linear.x)
+            temp.append(p)
+        return temp
 
     def pose_cb(self, msg):
-        # Called at 50 Hz
+        # TODO: Implement
         self.pose = msg
 
-
-    # waypoints is single-shot, whole track (Lane), so save away for later use, optimized for searching
     def waypoints_cb(self, waypoints):
-        self.base_waypoints = waypoints
-        # make sure self.waypoints_2d is initialized before the subscriber is
+        # TODO: Implement
+        self.base_lane = waypoints
         if not self.waypoints_2d:
-            self.waypoints_2d = [[waypoint.pose.pose.position.x, waypoint.pose.pose.position.y]
-                                 for waypoint in waypoints.waypoints]
+            self.waypoints_2d = [[waypoint.pose.pose.position.x,waypoint.pose.pose.position.y] for waypoint in waypoints.waypoint]
```

Comment:
At the end it should be `in waypoints.waypoints]`, it was off the screen in the walkthrough! ;-)